lib/Dialect/SV/Transforms/HWSpecialize.cpp

```diff
@@ -0,0 +1,298 @@
+    rewriter.replaceOpWithNewOp<ArrayGetOp>(op, op.input(), narrowedIndex);
+    return success();
+  }
+};
+
+// Generic pattern to convert parametric result types.
+struct ParametricTypeConversionPattern : public ConversionPattern {
+  ParametricTypeConversionPattern(MLIRContext *ctx,
+                                  TypeConverter &typeConverter,
+                                  ArrayAttr parameters)
+      : ConversionPattern(typeConverter, MatchAnyOpTypeTag(), /*benefit=*/1,
+                          ctx),
+        parameters(parameters) {}
+
+  LogicalResult
+  matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+                  ConversionPatternRewriter &rewriter) const override {
+    llvm::SmallVector<Value, 4> convertedOperands;
+    // Update the result types of the operation
+    bool ok = true;
+    rewriter.updateRootInPlace(op, [&]() {
+      // Mutate result types
+      for (auto &it : llvm::enumerate(op->getResultTypes())) {
+        auto res = evaluateParametricType(op->getLoc(), parameters, it.value());
+        ok &= succeeded(res);
+        if (!ok)
+          return;
+        op->getResult(it.index()).setType(res.getValue());
+      }
+    });
+
+    // Note: 'operands' have already been converted with the supplied type
+    // converter to this pattern. Make sure that we materialize this conversion
+    // by updating the operands to op.
+    op->setOperands(operands);
+
+    return success(ok);
+  };
+  ArrayAttr parameters;
+};
+
+struct HWSpecializePass : public sv::HWSpecializeBase<HWSpecializePass> {
+  void runOnOperation() override;
+};
+
+static void populateTypeConversion(Location loc, TypeConverter &typeConverter,
+                                   ArrayAttr parameters) {
+  // Possibly parametric types
+  typeConverter.addConversion([=](hw::IntType type) {
+    return evaluateParametricType(loc, parameters, type).getValue();
+  });
+  typeConverter.addConversion([=](hw::ArrayType type) {
+    return evaluateParametricType(loc, parameters, type).getValue();
+  });
+
+  // Valid target types.
+  typeConverter.addConversion([](mlir::IntegerType type) { return type; });
+}
+
+// Specializes the provided 'base' module into the 'target' module. By doing
+// so, we create a new module which
+// 1. has no parameters
+// 2. has a name composing the name of 'base' as well as the 'parameters'
+// parameters.
+// 3. Has a top-level interface with any parametric types resolved.
+// 4. Any references to module parameters have been replaced with the
+// parameter value.
+static LogicalResult specializeModule(OpBuilder builder, ArrayAttr parameters,
+                                      HWModuleOp source, HWModuleOp &target) {
+  auto *ctx = builder.getContext();
+  // Update the types of the source module ports based on evaluating any
+  // parametric in/output ports.
+  auto ports = source.getPorts();
+  for (auto &in : llvm::enumerate(source.getFunctionType().getInputs())) {
+    auto resType =
```

Comment:
nit 1: plz avoid using `auto` if it's not obvious from the call what the return type is.
supernit 2: I read `resType` as operation result type. I assume it's short for resolved type?